hr_timesheet_role/models/account_analytic_line.py

```diff
@@ -82,24 +73,27 @@ def _is_role_valid(self):
         if not self.project_id or not user_id:  # pragma: no cover
             return False
 
-        return (self.env['project.assignment'].search_count([
-            ('project_id', '=', self.project_id.id),
-            ('role_id', '=', self.role_id.id),
-            ('user_id', '=', user_id.id),
-        ]) > 0)
+        ProjectRole = self.env['project.role']
+        return bool(ProjectRole.search(
+            expression.AND([
+                [('id', '=', self.role_id.id)],
```

Comment:
Looking for an specific ID already reduces the search to one record, isn't it?

Reply:
Yes, which may or may not be allowed by other domain. Basically, I'm offloading filtering to database

Reply:
Then I don't get this search. Just do `bool(ProjectRole.browse(self.role_id.id).exists())`

Reply:
It's making `AND` of 2 domains, not just asking for a specific ID. Same as 
`self.role_id in ProjectRole.search(other_domain)`
What you propose is not the same

Reply:
And AND of something very restrictive (one record) and other thing gets nothing or that record, so I don't get why doing this way with such obscure code.

Reply:
Other domain can return more than that record. It's `self.role_id in ProjectRole.search(other_domain)` shifted to database layer

Reply:
Impossible with an AND...

Reply:
Why? What would be result of
```
ProjectRole.search([
    '&',
    ('id', '=', 45),
    ('id', 'in', [27, 28]),
])
```
vs
```
ProjectRole.search([
    '&',
    ('id', '=', 45),
    ('id', 'in', [27, 28, 45]),
])
```
?

Reply:
I don't really get what do you mean by impossible, for the queries above

Reply:
First will return nothing and the second only 45

Reply:
Exactly what I need. Replacement of
```
roles = ProjectRole.search([('id', 'in', [27, 28, 45])])
return self.role_id in roles
```
to avoid reading unknown amount records from database

Reply:
But there's no read in that code... The read is only done when you access other field apart from ID. Any way, what you have to do is:
```python
self.role_id.id in ProjectRole.get_available_roles_domain(
    user_id,
    self.project_id
)[2]
```
for clarity and for avoiding extra queries (the select one performed in the search)

Reply:
No, that is a huge no. There's no guarantee that get_available_roles_domain would return `[('id', 'in', [27, 28, 45])]`, it may as well be `[('company_id', '=', 1)]`

Reply:
And that's super-bad for extensibility. Right now I don't understand arguments against this piece of code.

Reply:
Yeah, but that's because you haven't planned correctly your API. A method `get_available_roles` or similar is mandatory for avoiding these "hacks" through the domain. Calling that method is the ideal, but I tried to adapt to what I knew:
```python
self.role_id.id in ProjectRole.get_available_roles(
    user_id,
    self.project_id
)
```
`get_available_roles_domain` must call this one as well and compose the rest.

Reply:
The problem with this code is that it's very bad in readability and less performant.

Reply:
There are three approaches:
A.
`get_available_roles_domain()` calls `get_available_roles()` and returns literally `[('id', 'in', get_available_roles().ids)]`

B.
`get_available_roles()` calls `get_available_roles_domain()` and returns literally `ProjectRole.browse(get_available_roles_domain())`

C.
`get_available_roles()` and `get_available_roles_domain()` are implemented separately, probably causing business logic duplication.

What are your points that A is better than B?

Reply:
A is better than B because:

* It's better for extensibility: you return a recordset obtain from any thing. The other way, you need to transpose a possible recordset in a search of type `('id', 'in', recordset.ids)`, which may be performant.
* But it's not only potentially less performant, but requires double search (the first with the real query), and the second with the ids.

Reply:
> bad in readability

Well, I'd argue on this why making valid boolean expressions on filters is bad for readability?

> less performant

This is simply not true, most of complexity is done on database side, while for cases of domain being `[('company_id', '=', 42)]` that could potentially return 50 roles, current implementation is more performant that selecting 50 roles and then filtering them on Odoo side

Reply:
This debate is not leading anywhere. What is your suggestion on making it according to your vision?

Reply:
I have already told in https://github.com/OCA/timesheet/pull/281#discussion_r343718471 you based on my experience at long term.

Reply:
> get_available_roles_domain must call this one as well and compose the rest.

Are you ok that for "all roles" case, it's going to be "[('id', 'in', [.....very many ids....])]" ?

Reply:
I'm not sure where do you need to return that domain instead the recordset itself. The only valid case that I see is on a domain on a m2o/m2m, which by other way is pretty the only way to return a valid domain.

Reply:
https://github.com/OCA/timesheet/pull/281/files#diff-76db791c32978da1ccc651215dbb49f3R88
https://github.com/OCA/timesheet/blob/12.0/hr_timesheet_role/models/account_analytic_line.py#L47

This is the only use-place of domain ^

Reply:
Yeah, what I supposed... There you shouldn't use other domain different from `('id', 'in...`, or you have the risk that the referred field is not included in the view and then you get an error.

Reply:
https://github.com/OCA/project/pull/607

Reply:
Updated